File: security-platform/agents/node/src/policy.ts
import { PolicyRule } from './types';

export class PolicyEngine {
  private mode: 'observe' | 'block';
  private rules: PolicyRule[];

  constructor(config: { mode?: 'observe' | 'block'; rules?: PolicyRule[] }) {
    this.mode = config.mode || 'observe';
    this.rules = config.rules || [];
  }

  check(action: string, context: Record<string, any>): { allowed: boolean; reason?: string } {
    // In observe mode, always allow but log
    if (this.mode === 'observe') {
      return { allowed: true };
    }

    // Check rules
    for (const rule of this.rules) {
      if (rule.action === action || rule.action === '*') {
        if (this.evaluateCondition(rule.condition, context)) {
          return {
            allowed: rule.effect === 'allow',
            reason: rule.id,
          };
        }
      }
    }

    // Default allow if no rules match
    return { allowed: true };
  }

  private evaluateCondition(condition: string, context: Record<string, any>): boolean {
    // Simple JSONPath-like evaluation
    // In real implementation, use a proper expression evaluator
    try {
      // For now, simple key-value matching
      const parts = condition.split('==');
      if (parts.length === 2) {
        const key = parts[0].trim();
        const value = parts[1].trim().replace(/['"]/g, '');
        return context[key] === value;
      }
      return false;
    } catch (e) {
      return false;
    }
  }
}

File: security-platform/agents/go/policy.go
package agent

import (
	"fmt"
	"strings"
)

// PolicyEngine enforces local policies.
type PolicyEngine struct {
	config *PolicyConfig
}

// NewPolicyEngine creates a new policy engine.
func NewPolicyEngine(config *PolicyConfig) *PolicyEngine {
	if config == nil {
		config = &PolicyConfig{Mode: "observe"}
	}
	return &PolicyEngine{config: config}
}

// Check checks if an action is allowed by policy.
func (p *PolicyEngine) Check(action string, context map[string]interface{}) *PolicyResult {
	mode := p.config.Mode
	if mode == "" {
		mode = "observe"
	}

	if strings.ToLower(mode) == "block" {
		rules := p.config.Rules
		if rules != nil {
			for _, rule := range rules {
				ruleAction, ok := rule["action"].(string)
				if !ok {
					continue
				}

				if strings.EqualFold(ruleAction, action) {
					effect, ok := rule["effect"].(string)
					if ok && strings.ToLower(effect) == "deny" {
						ruleID := "unknown"
						if id, ok := rule["id"].(string); ok {
							ruleID = id
						}
						return &PolicyResult{
							Allowed: false,
							Reason:  fmt.Sprintf("Blocked by policy rule: %s", ruleID),
						}
					}
				}
			}
		}
		return &PolicyResult{Allowed: true}
	}

	return &PolicyResult{
		Allowed: true,
		Reason:  "Observe mode - action allowed",
	}
}
